From be1a0dc709417cb3a81ef674123a25f80d55ce1a Mon Sep 17 00:00:00 2001
From: Alexandru Croitor <alexandru.croitor@qt.io>
Date: Wed, 10 Sep 2025 11:44:40 +0200
Subject: [PATCH 123/553] CMake: Add an SBOM LICENSE_DIR_PATHS project option

To allow specifying additional license directories via an option,
rather than via setting a QT_SBOM_LICENSE_DIRS variable in the parent
scope.

The old variable approach is kept for compatibility.

Pick-to: 6.9 6.8
Task-number: QTBUG-134894
Change-Id: Idae96c719dbc168b3366a0206c19393e27c7f76d
Reviewed-by: Joerg Bornemann <joerg.bornemann@qt.io>
(cherry picked from commit 3c57c7fed6a5f65a8a7ebe677f841241eaace9e8)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 cmake/QtPublicSbomHelpers.cmake               | 20 ++++++++++++++-----
 .../cmake/RunCMake/Sbom/common_targets.cmake  |  6 ++++++
 .../LicenseRef-TestLicense.txt                |  1 +
 tests/auto/cmake/RunCMake/Sbom/full.cmake     |  3 +++
 4 files changed, 25 insertions(+), 5 deletions(-)
 create mode 100644 tests/auto/cmake/RunCMake/Sbom/custom_licenses/LicenseRef-TestLicense.txt

diff --git a/cmake/QtPublicSbomHelpers.cmake b/cmake/QtPublicSbomHelpers.cmake
index c5d54ed199e..25c6c52b1b1 100644
--- a/cmake/QtPublicSbomHelpers.cmake
+++ b/cmake/QtPublicSbomHelpers.cmake
@@ -47,6 +47,7 @@ function(_qt_internal_sbom_begin_project)
     )
     set(multi_args
         COPYRIGHTS
+        LICENSE_DIR_PATHS
     )
 
     cmake_parse_arguments(PARSE_ARGV 0 arg "${opt_args}" "${single_args}" "${multi_args}")
@@ -285,15 +286,24 @@ function(_qt_internal_sbom_begin_project)
         list(APPEND license_dirs "${PROJECT_SOURCE_DIR}/LICENSES")
     endif()
 
+    set(license_dir_candidates "")
+    if(arg_LICENSE_DIR_PATHS)
+        list(APPEND license_dir_candidates ${arg_LICENSE_DIR_PATHS})
+    endif()
+
     # Allow specifying extra license dirs via a variable. Useful for standalone projects
     # like sqldrivers.
+    # Kept for backwards compatibility, the new LICENSE_DIR_PATHS option should be preferred.
     if(QT_SBOM_LICENSE_DIRS)
-        foreach(license_dir IN LISTS QT_SBOM_LICENSE_DIRS)
-            if(EXISTS "${license_dir}")
-                list(APPEND license_dirs "${license_dir}")
-            endif()
-        endforeach()
+        list(APPEND license_dir_candidates ${QT_SBOM_LICENSE_DIRS})
     endif()
+
+    foreach(license_dir IN LISTS license_dir_candidates)
+        if(EXISTS "${license_dir}")
+            list(APPEND license_dirs "${license_dir}")
+        endif()
+    endforeach()
+
     list(REMOVE_DUPLICATES license_dirs)
 
     set(license_file_wildcard "LicenseRef-*.txt")
diff --git a/tests/auto/cmake/RunCMake/Sbom/common_targets.cmake b/tests/auto/cmake/RunCMake/Sbom/common_targets.cmake
index bde4ab96ee1..ce2f39de3d6 100644
--- a/tests/auto/cmake/RunCMake/Sbom/common_targets.cmake
+++ b/tests/auto/cmake/RunCMake/Sbom/common_targets.cmake
@@ -47,6 +47,12 @@ _qt_internal_add_sbom(app
     RUNTIME_PATH bin
 )
 
+if(IS_FULL_BUILD)
+    _qt_internal_extend_sbom(app
+        LICENSE_EXPRESSION "LicenseRef-TestLicense"
+    )
+endif()
+
 find_package(ZLIB)
 if(ZLIB_FOUND)
     _qt_internal_add_sbom(ZLIB::ZLIB
diff --git a/tests/auto/cmake/RunCMake/Sbom/custom_licenses/LicenseRef-TestLicense.txt b/tests/auto/cmake/RunCMake/Sbom/custom_licenses/LicenseRef-TestLicense.txt
new file mode 100644
index 00000000000..b12d29cb1e8
--- /dev/null
+++ b/tests/auto/cmake/RunCMake/Sbom/custom_licenses/LicenseRef-TestLicense.txt
@@ -0,0 +1 @@
+test license contents
diff --git a/tests/auto/cmake/RunCMake/Sbom/full.cmake b/tests/auto/cmake/RunCMake/Sbom/full.cmake
index 35dd2a2bdfb..7241b8e77a8 100644
--- a/tests/auto/cmake/RunCMake/Sbom/full.cmake
+++ b/tests/auto/cmake/RunCMake/Sbom/full.cmake
@@ -5,6 +5,8 @@ _qt_internal_setup_sbom(
     GENERATE_SBOM_DEFAULT "TRUE"
 )
 
+set(IS_FULL_BUILD "TRUE")
+
 _qt_internal_sbom_begin_project(
     SBOM_PROJECT_NAME "${PROJECT_NAME}Project"
     SUPPLIER "QtProjectTest"
@@ -17,6 +19,7 @@ _qt_internal_sbom_begin_project(
     CPE "cpe:2.3:a:qt:qtprojecttest:1.0.0:*:*:*:*:*:*:*"
     VERSION "1.0.0"
     DOCUMENT_CREATOR_TOOL "Test Build System Tool"
+    LICENSE_DIR_PATHS "${CMAKE_CURRENT_SOURCE_DIR}/custom_licenses"
 )
 
 include(common_targets.cmake)
-- 
2.51.2

